add new tech and demos. (#11)
ScreaMy7/owasp-mastg: techniques/android/MASTG-TECH-XXXX.md

---
title: Deep Link Testing
platform: android
---

This technique groups common tasks for assessing deep links on Android: detecting usage, validating website association, monitoring its resolution, and invoking links.

> ## Check for Deep Link Usage

Reviewer: ```suggestion
## Verify Usage of Deep Links
```

> Identify deep link usage via manifest inspection and code review.

Reviewer: ```suggestion
Deep link entry points can be discovered by examining intent filters in the `AndroidManifest.xml` and reviewing the code to understand how incoming deep link data is processed and validated.
```

> ### Manifest inspection

Reviewer: ```suggestion
### Inspection of the AndroidManifest File
```

Extract and review intent-filters that declare VIEW/BROWSABLE handlers. Use @MASTG-TOOL-0124, then select `intent-filter` with `android.intent.action.VIEW"` and `android.intent.category.BROWSABLE`.

```bash
aapt2 dump xmltree app.apk AndroidManifest.xml 
```

> ### Using Dumpsys:

Reviewer: ```suggestion
### Using Dumpsys
```

> Use  adb to run the following command that will show all schemes:

Reviewer: ```suggestion
Use adb to run the following command that will show all schemes
```

`adb shell dumpsys package com.example.package`

> ### Using Android "App Link Verification" Tester:

Reviewer: ```suggestion
### Using the Android App Link Verification Tester
```

Use the Android @MASTG-TOOL-014X Tester to list all deep links (list-all) or only app links (list-applinks):

```bash
$ python3 deeplink_analyser.py -op list-all -apk example.apk

org.owasp.mastestapp.MainActivity

vulnerable-app://deeplink
```

> ## Check for Correct Website Association

Reviewer: ```suggestion
## Verify Domain Association Setup for App Links
```

> Android App Links must be verified against a website’s Digital Asset Links to ensure only the legitimate app handles links for that domain.

Reviewer: ```suggestion
Android App Links must be verified against a website's Digital Asset Links to ensure only the legitimate app handles links for that domain.
```

> ### Verify App Links state 

Reviewer: ```suggestion
### Verify App Links state
```

```bash
# Trigger verification
adb shell pm verify-app-links com.example.app

# Inspect current state
adb shell pm get-app-links com.example.app

# Reset selection (optional) and re-verify
adb shell pm reset-app-links com.example.app
adb shell pm verify-app-links com.example.app
```

Preferential handling and verification status per host should be listed. Unverified hosts indicate missing or incorrect associations.

### Validate assetlinks.json

> Fetch the site statements and verify package name and sha256_cert_fingerprints match the app’s signing cert:

Reviewer: ```suggestion
Fetch the site statements and verify package name and sha256_cert_fingerprints match the app's signing cert
```

```bash
curl -s https://example.com/.well-known/assetlinks.json | jq
```

Check for invalid Digital Asset Links files served via HTTPS. For example:
- The file contains invalid JSON.
- The file doesn't include the target app's package.
- If an intent filter lists multiple hosts with different subdomains, there must be a valid Digital Asset Links file on each domain.

@MASTG-TOOL-014X can be also used get the verification status for all app links (verify-applinks).

## Monitoring Deep Links

When using deep links, monitor how the system resolves and dispatches the Intent.

> ### Logcat

Reviewer: ```suggestion
###  Using Logcat
```
